propolis/src/block.rs

```diff
@@ -108,69 +194,126 @@ impl<R: BlockReq> PlainBdev<R> {
             }
 
             if let Some(mut req) = reqs.pop_front() {
-                let res = match req.oper() {
-                    BlockOp::Flush => {
-                        let res =
-                            unsafe { libc::fdatasync(self.fp.as_raw_fd()) };
-                        if res == -1 {
-                            BlockResult::Failure
-                        } else {
-                            BlockResult::Success
-                        }
-                    }
-                    BlockOp::Read => self.process_read(&mut req, ctx),
-                    BlockOp::Write => self.process_write(&mut req, ctx),
-                };
-                req.complete(res, ctx);
+                let result = self.process_request(&mut req, ctx);
+                req.complete(result, ctx);
             } else {
                 reqs = self.cond.wait(reqs).unwrap();
             }
         }
     }
-    fn process_read(&self, req: &mut R, ctx: &DispCtx) -> BlockResult {
+
+    /*
+     * Match against individual BlockReqs and call appropriate processing functions.
+     */
+    fn process_request(&self, req: &mut R, ctx: &DispCtx) -> BlockResult {
         let mem = ctx.mctx.memctx();
 
         let mut offset = req.offset();
+
+        // TODO: are buffers guaranteed to be sector size? do BlockDevBackingStores require
+        // this?
         while let Some(buf) = req.next_buf() {
-            if let Some(mapping) = mem.writable_region(&buf) {
-                if let Ok(nread) = mapping.pread(&self.fp, buf.1, offset as i64)
-                {
-                    assert_eq!(nread as usize, buf.1);
-                    offset += buf.1;
-                } else {
-                    // XXX: Error reporting (bad read)
+            let sz = buf.1;
+
+            let result = match req.oper() {
+                BlockOp::Read => {
+                    self.process_read_request(offset, sz, &mem, buf)
+                }
+                BlockOp::Write => self.process_write_request(offset, &mem, buf),
+                BlockOp::Flush => self.process_flush(),
+            };
+
+            // TODO: should all BlockReq be attempted instead of returning early?
```

Comment:
I'm inclined to say if we fail to read/write/flush one of the BlockReqs, we should abort the IO request and signal as such to the guest.

Reply:
Yeah, I agree. Fixed in 234e243